File: samples/kmp/androidApp/src/main/kotlin/jp/kamusoft/ksdialogs/samples/kmp/android/SampleToastRegistration.kt
package jp.kamusoft.ksdialogs.samples.kmp.android

import jp.kamusoft.ksdialogs.Toast
import jp.kamusoft.ksdialogs.samples.kmp.CustomToastViewModel

/**
 * 共有コードのカスタム Toast の ViewModel 型と Android の View factory の紐付け。
 *
 * Android では共有コードの ViewModel 契約が Native ライブラリのものと同一の型になるため、
 * 登録は Native の登録面へそのまま行える (kmp/ADR-0002)。
 * Toast のレジストリは Dialog / Loading のものとは独立している。
 */
internal object SampleToastRegistration {
    /** この Sample が使うカスタム Toast を登録する。 */
    fun register() {
        // 中身は ViewModel だけを受け取るので、(Context, VM) を受けるコンストラクタの参照をそのまま渡せる
        Toast.instance.registry.register(CustomToastViewModel::class, ::CustomToastCardView)
    }
}

File: samples/kmp/iosApp/KsDialogsSampleKmp/SampleToastRegistration.swift
import KsDialogs
import SampleShared

/// 共有コードのカスタム Toast の ViewModel 型と iOS の View factory の紐付け。
///
/// 共有コードの ViewModel は Swift の ViewModel 契約に準拠しないため、登録は KMP 面 (`kmp`) へ行う。
/// 共有コードで定義した ViewModel のクラスがそのまま登録キーになり、
/// 共有 Presenter からの表示と同じレジストリを引く。
enum SampleToastRegistration {
    /// この Sample が使うカスタム Toast を登録する。
    @MainActor
    static func register() {
        // SwiftUI の View をそのまま返す登録。UIView を返す登録と同じ名前で、戻り値の型だけが違う
        Toast.shared.kmp.register(CustomToastViewModel.self) { viewModel in
            CustomToastCard(viewModel: viewModel)
        }
    }
}
